pkg/storage/ingest/pusher.go

```diff
@@ -65,51 +66,67 @@ func newPusherConsumer(p Pusher, reg prometheus.Registerer, l log.Logger) *pushe
 }
 
 func (c pusherConsumer) consume(ctx context.Context, records []record) error {
-	recC := make(chan parsedRecord)
 	ctx, cancel := context.WithCancelCause(ctx)
 	defer cancel(cancellation.NewErrorf("done consuming records"))
 
+	recC := make(chan parsedRecord)
+
 	// Speed up consumption by unmarhsalling the next request while the previous one is being pushed.
 	go c.unmarshalRequests(ctx, records, recC)
-	err := c.pushRequests(ctx, recC)
+	err := c.pushRequests(recC)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-func (c pusherConsumer) pushRequests(ctx context.Context, reqC <-chan parsedRecord) error {
+func (c pusherConsumer) pushRequests(reqC <-chan parsedRecord) error {
 	recordIdx := -1
 	for wr := range reqC {
 		recordIdx++
 		if wr.err != nil {
 			level.Error(c.l).Log("msg", "failed to parse write request; skipping", "err", wr.err)
 			continue
 		}
-		processingStart := time.Now()
 
-		ctx := user.InjectOrgID(ctx, wr.tenantID)
-		err := c.p.PushToStorage(ctx, wr.WriteRequest)
+		ctx := user.InjectOrgID(wr.ctx, wr.tenantID)
+		err := c.pushToStorage(ctx, wr.WriteRequest)
+		if err != nil {
+			return fmt.Errorf("consuming record at index %d for tenant %s: %w", recordIdx, wr.tenantID, err)
+		}
+	}
+	return nil
+}
+
+func (c pusherConsumer) pushToStorage(ctx context.Context, req *mimirpb.WriteRequest) error {
+	spanLog, spanCtx := spanlogger.NewWithLogger(ctx, c.l, "pusherConsumer.pushToStorage")
+	defer spanLog.Finish()
 
-		c.processingTimeSeconds.Observe(time.Since(processingStart).Seconds())
-		c.totalRequests.Inc()
+	processingStart := time.Now()
 
-		if err != nil {
-			if !mimirpb.IsClientError(err) {
-				c.serverErrRequests.Inc()
-				return fmt.Errorf("consuming record at index %d for tenant %s: %w", recordIdx, wr.tenantID, err)
-			}
-			c.clientErrRequests.Inc()
+	err := c.p.PushToStorage(spanCtx, req)
+
+	c.processingTimeSeconds.Observe(time.Since(processingStart).Seconds())
+	c.totalRequests.Inc()
 
-			// The error could be sampled or marked to be skipped in logs, so we check whether it should be
-			// logged before doing it.
-			if keep, reason := shouldLog(ctx, err); keep {
-				if reason != "" {
-					err = fmt.Errorf("%w (%s)", err, reason)
-				}
+	if err != nil {
+		// Only return non-client errors; these will stop the processing of the current Kafka fetches and retry (possibly).
+		if !mimirpb.IsClientError(err) {
+			c.serverErrRequests.Inc()
+			return spanLog.Error(err)
+		}
 
-				level.Warn(c.l).Log("msg", "detected a client error while ingesting write request (the request may have been partially ingested)", "err", err, "user", wr.tenantID)
+		c.clientErrRequests.Inc()
+
+		// The error could be sampled or marked to be skipped in logs, so we check whether it should be
+		// logged before doing it.
+		if keep, reason := shouldLog(ctx, err); keep {
+			if reason != "" {
+				err = fmt.Errorf("%w (%s)", err, reason)
 			}
+
+			tenantID, _ := user.ExtractOrgID(spanCtx)
```

Comment:
Can we take the tenantID directly in input in `pushToStorage()` instead of extracting it?